components/lcd/esp_lcd_sh1107/esp_lcd_sh1107.c

```diff
@@ -0,0 +1,326 @@
+    sh1107_panel_t *sh1107 = NULL;
+    ESP_GOTO_ON_FALSE(io && panel_dev_config && ret_panel, ESP_ERR_INVALID_ARG, err, TAG, "invalid argument");
+    ESP_GOTO_ON_FALSE(panel_dev_config->bits_per_pixel == 1, ESP_ERR_INVALID_ARG, err, TAG, "bpp must be 1");
+    esp_lcd_panel_sh1107_config_t *vendor_cfg = (esp_lcd_panel_sh1107_config_t *) panel_dev_config->vendor_config;
+    ESP_GOTO_ON_FALSE(vendor_cfg, ESP_ERR_INVALID_ARG, err, TAG, "vendor config cannot be null");
+    sh1107 = calloc(1, sizeof(sh1107_panel_t));
+    ESP_GOTO_ON_FALSE(sh1107, ESP_ERR_NO_MEM, err, TAG, "no mem for sh1107 panel");
+
+    if (GPIO_IS_VALID_OUTPUT_GPIO(panel_dev_config->reset_gpio_num)) {
+        gpio_config_t io_conf = {
+            .mode = GPIO_MODE_OUTPUT,
+            .pin_bit_mask = 1ULL << panel_dev_config->reset_gpio_num,
+        };
+        ESP_GOTO_ON_ERROR(gpio_config(&io_conf), err, TAG, "configure GPIO for RST line failed");
+    }
+
+    sh1107->io = io;
+    sh1107->lcd_width = vendor_cfg->lcd_width;
+    sh1107->lcd_height = vendor_cfg->lcd_height;
+    sh1107->bits_per_pixel = panel_dev_config->bits_per_pixel;
+    sh1107->reset_gpio_num = panel_dev_config->reset_gpio_num;
+    sh1107->reset_level = panel_dev_config->flags.reset_active_high;
+    sh1107->base.del = panel_sh1107_del;
+    sh1107->base.reset = panel_sh1107_reset;
+    sh1107->base.init = panel_sh1107_init;
+    sh1107->base.draw_bitmap = panel_sh1107_draw_bitmap;
+    sh1107->base.invert_color = panel_sh1107_invert_color;
+    sh1107->base.set_gap = panel_sh1107_set_gap;
+    sh1107->base.mirror = panel_sh1107_mirror;
+    sh1107->base.swap_xy = panel_sh1107_swap_xy;
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 0, 0)
+    sh1107->base.disp_off = panel_sh1107_disp_on_off;
+#else
+    sh1107->base.disp_on_off = panel_sh1107_disp_on_off;
+#endif
+    *ret_panel = &(sh1107->base);
+    ESP_LOGD(TAG, "new sh1107 panel @%p", sh1107);
+
+    return ESP_OK;
+
+err:
+    if (sh1107) {
+        if (GPIO_IS_VALID_OUTPUT_GPIO(panel_dev_config->reset_gpio_num)) {
+            gpio_reset_pin(panel_dev_config->reset_gpio_num);
+        }
+        free(sh1107);
+    }
+    return ret;
+}
+
+static esp_err_t panel_sh1107_del(esp_lcd_panel_t *panel)
+{
+    sh1107_panel_t *sh1107 = __containerof(panel, sh1107_panel_t, base);
+    if (GPIO_IS_VALID_OUTPUT_GPIO(sh1107->reset_gpio_num)) {
+        gpio_reset_pin(sh1107->reset_gpio_num);
+    }
+    ESP_LOGD(TAG, "del sh1107 panel @%p", sh1107);
+    free(sh1107);
+    return ESP_OK;
+}
+
+static esp_err_t panel_sh1107_reset(esp_lcd_panel_t *panel)
+{
+    sh1107_panel_t *sh1107 = __containerof(panel, sh1107_panel_t, base);
+
+    // perform hardware reset
+    if (GPIO_IS_VALID_OUTPUT_GPIO(sh1107->reset_gpio_num)) {
+        gpio_set_level(sh1107->reset_gpio_num, sh1107->reset_level);
+        vTaskDelay(pdMS_TO_TICKS(10));
+        gpio_set_level(sh1107->reset_gpio_num, !sh1107->reset_level);
+        vTaskDelay(pdMS_TO_TICKS(10));
+    }
+
```

Comment:
Could we add software reset too?

Reply:
- There isn't software reset.
- About selection I2C address - is it good to put selection into KConfig?